### PR TITLE
Suppress all C++ warning when compiling generated protobuf code.

### PR DESCRIPTION
**Pull request by RNabel**
This PR is a suggested fix for #5466. 

As long as a) one cannot specify `copt`s for the compilation of generated C++ code, and b) protobuf generated code is warning-generating, we should suppress _at least_ the errors protobuf code is known to generate (from my experience at least `unused-parameter` (as seen here: https://github.com/google/protobuf/issues/4628) and `invalid-offset` and `zero-as-null-pointer-constant`). 

a) is fixable, but has no timeline, b) will not be fixed.

## Option 1: Suppress all warnings (currently chosen)
Since protobuf code is already assumed to be correct, suppressing all warnings both reduces log message spam and solves the `-Werror` problem.
**Advantages**: Will compile with any CROSSTOOL copts.
**Disadvantages**: Could suppress warnings of a broken C++ protobuf `plugin`.

## Option 2: Suppress specific known errors 
It is theoretically possible to suppress only the known errors, but I'm not a C++ expert so don't know how compatible these flags are between OSs.
**Risks**: Windows, different protobuf versions could have different errors.

## Option 3: Pass `-Wno-error`
Warnings would still be shown but at least the compilation would succeed.
**Advantages**: least intrusive, builds complete.
**Disadvantages**: many un-actionable log messages.

I hope this is the correct format for this kind of discussion. Many thanks in advance for any comments :)

### CODE REVIEW

**Comment by rongjiecomputer** [comment]
> It is theoretically possible to suppress only the known errors, but I'm not a C++ expert so don't know how compatible these flags are between OSs.
Risks: Windows, different protobuf versions could have different errors.

Context: I am C++ programmer on Windows and regularly deal with GCC, Clang and MSVC.

Adding only the flags needed for each compiler (the problem is specific to compiler, not OS) is feasible, but adding them in `.java` is kind of awkward. If `cc_proto_library` rule will be migrated to Skylark and flags can be added with Crosstool in Skylark, that will be easier for me.

`zero-as-null-pointer-constant` warning should be fixed in upstream protobuf in future version when they migrate from `NULL` to `nullptr` (they couldn't do it before 3.6.0 due to requirement to support pre-C++11 compilers).

File: src/main/java/com/google/devtools/build/lib/rules/cpp/proto/CcProtoAspect.java [hunk 1]
```diff
@@ -256,6 +256,7 @@ private CcCompilationHelper initializeCompilationHelper(
       }
 
       helper.addDeps(ruleContext.getPrerequisites("deps", TARGET));
+      helper.setCopts(ImmutableList.of("-w"));
```
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/rules/cpp/proto/CcProtoAspect.java]: This won't work: this assumes that the compiled is gcc/clang and leaves MSVC users out in the cold.

Comment (RNabel) [reply src/main/java/com/google/devtools/build/lib/rules/cpp/proto/CcProtoAspect.java]: Thanks for the review! I'll try it on Windows just to be on the safe side. 
The MSVC docs state [`Options are specified by either a forward slash (/) or a dash (–).`](https://msdn.microsoft.com/en-us/library/610ecb4h.aspx) (see the `option` entry). This in combination with [the Warning levels page](https://docs.microsoft.com/en-gb/cpp/build/reference/compiler-option-warning-level) leads me to believe that it _should_ work. But then again, this is Windows we're talking about.

Comment (rongjiecomputer) [reply src/main/java/com/google/devtools/build/lib/rules/cpp/proto/CcProtoAspect.java]: MSVC can accept `-w` to disable all warnings just fine, but if someone *append* `-W3`/`-W4` etc. *after* `-w`, you will get [D9025` warning](https://msdn.microsoft.com/en-us/library/8k3f51f1.aspx) that cannot be silenced at all, causing a waterfall of warnings.

**Comment by lberki** [comment]
Well, the review latency was embarrassing....

Even if it does work with MSVC, I'd like us to move away from hard-coded command line options not towards them. What do you think, @mhlopko ?

**Comment by RNabel** [comment]
@lberki I agree that having hard-coded command line options is non-optimal and that it would be _far_ better to add a feature to the `CROSSTOOL`s which suppresses all warnings and enable that instead. Am I correct in assuming that this `feature` would need to be added to the [general `CROSSTOOL`](https://github.com/bazelbuild/bazel/blob/master/tools/cpp/CROSSTOOL) and the [osx `CROSSTOOL`](https://github.com/bazelbuild/bazel/blob/master/tools/cpp/CROSSTOOL). How does the latter get edited? It states:
```
# This file was auto-generated by a script maintained internally by
# Google.
```

**Comment by mhlopko** [comment]
+1 what @lberki said, let's not hardcode any flags if we can help it. I propose to add "disable_errors" feature to requested features [here](https://source.bazel.build/bazel/+/55d0c1c649d3117f59e3d5d7b79de12d6b090381:src/main/java/com/google/devtools/build/lib/rules/cpp/proto/CcProtoAspect.java;l=219).

Yes that means that we have to define the feature in all the crosstools. Well, we don't have to (bazel won't complain if the feature is not defined) but users will expect the feature to be there, so let's not let them down.

Osx crosstool is indeed generated by a tool that we didn't get to open source (and we will not because Crosstool in Skylark solves the problem better). Just edit one toolchain in that file and I'll take care of the generator once I import the PR.

**Comment by RNabel** [comment]
@mhlopko I just pushed a working example for macos and added the new feature to what I hope is the default Unix toolchain. I tested it on macos (sorry don't have a Linux machine handy right now). 

Could you let me know where else this feature should go? 

Finally, I'd propose to rename the feature "disable_warnings" or changing the flag from `"-w"` to `"-Wno-error"`.

**Comment by laurentlb** [comment]
ping @mhlopko

**Comment by RNabel** [comment]
ping @lberki

**Comment by mhlopko** [comment]
I like it, let me import it :) Thanks!

**Comment by mhlopko** [comment]
(Asking protobuf team for approval internally, will take some time).

**Comment by RNabel** [comment]
@mhlopko Awesome, thanks for the update!

**Comment by RNabel** [comment]
@mhlopko Do you have a rough timeline for this to go in?

**Comment by petemounce** [comment]
Any update?

**Comment by jin** [comment]
What is this status of this PR?

**Comment by lberki** [comment]
Wow, this was a while ago... I imported this change and asked @mhlopko to review it. What's the plan for Windows?

**Comment by jin** [comment]
Is this still importable?

**Comment by hlopko** [comment]
Hi all,

I'm very sorry for how long this PR took. So there were discussion with the protobuf team internally and the result is that protoc should not generate warning triggering code. I commented https://github.com/protocolbuffers/protobuf/issues/4628#issuecomment-470446302 and I expect that protobuf team will reopen that issue and drive this effort further. @RNabel, @rongjiecomputer, can you check in on https://github.com/protocolbuffers/protobuf/issues/4628#issuecomment-470446302 to help them understand the actual problems?

Thank you! And again, I apologize for this taking this long.

I'll now close this PR. But please shout if you disagree with what I wrote above and we'll continue the discussion here.

**Pull request closed without merging**